Add more documentation for ament_cmake

Martin-Idel opened the pull request:
This PR tries to improve the documentation for ament_cmake.
CMake is a vast ecosystem with syntax going back a long time, so there are many ways to do things.
Ament_cmake also adds quite a bit of functionality, especially with the ament_index and the resource_index. There is documentation in the form of source code doxygen and one design document, but that's it from what I could see.

To help other developers use ament_cmake, I compiled what I learned during one and a half year of working with ament_cmake. I tried to give it a structure to make it as useful as possible, but comments are very welcome. 

The guidelines for my writup were the following:
- document everything I wish I had known at the time I needed it except for those things that are very specific to an external library or which are not overly CMake-specific
- use modern target-based CMake when possible and when it doesn't conflict with the usage I have seen in ROS 2
- use best practices as I have learned them while using ament_cmake and via review comments from core maintainers

This is probably too detailed in some areas and not detailed enough in others. It is also not a "tutorial" although I try to give snippets whenever it is necessary to show how to actually use the ideas.
Therefore, comments and suggestions are especially welcome.

hidmic commented on source/Tutorials/Ament-CMake-Documentation.rst:
```diff
@@ -0,0 +1,498 @@
+
+- a layout of the marker file, which can be anything and could also be empty (this is true for instance for the "package" resource marking a ROS 2 package)
+
+For the RViz mesh resource, the corresponding choices were
```
@Martin-Idel missing colon at the end.

Martin-Idel commented on source/Tutorials/Ament-CMake-Documentation.rst:
```diff
@@ -0,0 +1,498 @@
+.. redirect-from::
+
+    Ament-CMake-Documentation
```
Thanks for letting me know. I removed it. I've never used ReST before and just mimicked other documentation files.

hidmic replied:
That's fine. It means we need better documentation for the `redirect-from` directive :)

Martin-Idel commented on source/Tutorials/Ament-CMake-Documentation.rst:
```diff
@@ -0,0 +1,498 @@
+Here, we assume that the folder "include" contains the headers which need to be exported.
+Note that it is not necessary to put all headers into a separate folder, only those that should be included by clients.
+
+Here is what's happening in the snippet above:
```
That was my original intention, but I have missed a whole section and quite a few others I think. Should be much better now.

Martin-Idel commented on source/Tutorials/Ament-CMake-Documentation.rst:
```diff
@@ -0,0 +1,498 @@
+
+ament_cmake is the build system for CMake based packages in ROS 2 (in particular, it will be used for most if not all C/C++ projects).
+It is a set of scripts enhancing CMake and adding convenience functionality for package authors.
+Knowing the basics of CMake will be very helpful.
```
Good idea, I added two: one to the official cmake documentation (version 3.5 as this is the version ROS 2 uses throughout) and one to the official tutorial.

Martin-Idel commented:
@hidmic: Thank you for your comments. I reworked the PR accordingly and also resolved some other inconsistencies. I also added an initial section about `ament_package()` to streamline the presentation a bit further.

What's definitely missing is a treatment of `amen_cmake_python` modules, but I never really used those. If you think it's very important to have them now, I can add them, but since this requires me reading a bit more about them, I'll wait for more feedback.

hidmic commented on source/Tutorials/Ament-CMake-Documentation.rst:
```diff
@@ -0,0 +1,533 @@
+The argument to ``project`` will be the package name and must be identical to the package name in the package.xml.
+
+The project setup is done by ``ament_package()`` and this call must occur exactly once per package.
+``ament_package()`` installs the package.xml and registers the package with the ament index so that it can be found by other packages using ``find_package``.
```
@Martin-Idel hmm, AFAIK `ament_package()` does install the `package.xml` but what enables `find_package()` calls downstream are the CMake config files, not the ament index (see [here](https://cmake.org/cmake/help/v3.0/manual/cmake-packages.7.html#using-packages)).

Martin-Idel replied:
You are right of course, but I didn't mean to claim that this was what installing the package.xml did. I clarified it.

hidmic commented on source/Tutorials/Ament-CMake-Documentation.rst:
```diff
@@ -0,0 +1,533 @@
+^^^^^^^^^^^^^^^^^^^
+
+There are two ways to link your packages against a new dependency.
+The recommended way in modern CMake is to only use targets, eporting and linking against them.
```
@Martin-Idel `eporting` --> `exporting`

hidmic commented on source/Tutorials/Ament-CMake-Documentation.rst:
```diff
@@ -0,0 +1,533 @@
+
+- For Visual Studio, the default ``W1`` warnings are kept
+
+- For GCC and Clang: ``-Wall, -Wextra, -Wpedantic`` are required and ``-Wshadow, -Werror`` are advisable (the latter makes warnings errors).
```
@Martin-Idel I'd remove the commas so that the reader can just copy & paste them.

hidmic commented on source/Tutorials/Ament-CMake-Documentation.rst:
```diff
@@ -0,0 +1,533 @@
+
+- a cmake style linter
+
+- an xml linter
```
@Martin-Idel I believe that the actual list of linters to be run depends on the `package.xml` (see [here](https://github.com/ament/ament_lint/blob/master/ament_lint_auto/doc/index.rst)).

Martin-Idel replied:
Right, what I didn't say is that I was thinking about using this together with `ament_lint_common` :). 
I clarified this.

hidmic commented on source/Tutorials/Ament-CMake-Documentation.rst:
```diff
@@ -0,0 +1,533 @@
+
+    ament_uncrustify --reformat <path_to_source_folders>
+
+If one of the linters doesn't work for now, it can also be disabled via setting ``<linter>_FOUND`` to ``TRUE``, e.g.:
```
@Martin-Idel the value of disabling some linters is arguable already, but this just looks like a hack. How did you come across this trick?

Martin-Idel replied:
This is a hack of course. The value that I see in this is to disable linters temporarily until they can be run (again). For instance, RViz cannot use the copyright linter and certain rosbag linters cannot use cppcheck. 

I came across it here lately: https://github.com/ros2/rosbag2/blob/49647575f758b9ffeab088d0fc3309ba7b5d51b6/rosbag2_transport/CMakeLists.txt#L84

However, you are right that it's highly debatable whether this has to be explained in a documentation/tutorial which also tries to lay out some of the better practices. I think it's better to leave it out.

hidmic commented on source/Tutorials/Ament-CMake-Documentation.rst:
```diff
@@ -0,0 +1,533 @@
+Testing
+^^^^^^^
+
+Ament contains CMake macros to simplify setting up GTests. Call
```
@Martin-Idel missing colon at the end.

dirk-thomas commented on source/Tutorials/Ament-CMake-Documentation.rst:
```diff
@@ -0,0 +1,530 @@
+
+A basic CMake outline can be produced using ``ros2 pkg create <package_name>`` on the command line.
+The basic build information is then gathered in two files: the ``package.xml`` and the ``CMakeLists.txt``.
+The ``package.xml`` must contain all dependencies and a bit of metadata to allow colcon/ament to find the correct build order for your packages.
```
> allow colcon/ament to find the correct build order

I would remove `/ament` here since it isn't involved in ordering packages.

dirk-thomas replied:
Maybe append the following to clarify that the information is not used only by the build tool:

> , to install the required dependencies in CI as well as provide the information for a release with ``bloom``.

dirk-thomas commented on source/Tutorials/Ament-CMake-Documentation.rst:
```diff
@@ -0,0 +1,530 @@
+  For an example of when to use these arguments, see the discussion in `Adding resources`_.
+  For more information on how to use template files, see `the official documentation <https://cmake.org/cmake/help/v3.5/command/configure_file.html>`__.
+
+- ``CONFIG_EXTRAS_POST``: same as ``CONFIG_EXTRAS``
```
"same" is misleading here. Obviously there is a difference otherwise the option wouldn't exist. Maybe clarify in which order the two are being included in regards to other exported information.

Martin-Idel replied:
Okay, I think I corrected it. Please have a look.

dirk-thomas commented on source/Tutorials/Ament-CMake-Documentation.rst:
```diff
@@ -0,0 +1,530 @@
+
+The following best practice is proposed:
+
+- if you are building a library, put all headers which should be usable by clients and therefore must be installed into and ``include`` folder, while all other files (``.c/.cpp`` and header files which should not be exported) are inside the ``src`` folder.
```
Explicitly mention that the all headers should be in a sundirectory of `include` named after the package.

dirk-thomas commented on source/Tutorials/Ament-CMake-Documentation.rst:
```diff
@@ -0,0 +1,530 @@
+This adds all files in the folder ``${CMAKE_CURRENT_SOURCE_DIR}/include`` to the public interface during build time and
+all files in the include folder (relative to ``${CMAKE_INSTALL_DIR}``) when being installed.
+
+In principle, this is not necessary if both folders are called ``include`` and top-level with respect to ``${CMAKE_CURRENT_SOURCE_DIR}`` and ``${CMAKE_INSTALL_DIR}``, but it is very common.
```
Why would this not be necessary? Without the build wouldn't find the headers.

Martin-Idel replied:
Good catch. I meant to say that using *generator expressions* is optional, not using `target_include_directories`. That's of course necessary.

dirk-thomas commented on source/Tutorials/Ament-CMake-Documentation.rst:
```diff
@@ -0,0 +1,530 @@
+CMake targets are namespaced, similar to C++. For instance, a target may be Eigen3::Eigen.
+
+If a target version of the dependency exists, it is advised to use this version.
+As an example, suppose we want to link ``my_library`` against the linear algebra library Eigen3. Then it suffices to call
```
NItpick: wrap after each sentence.

dirk-thomas commented on source/Tutorials/Ament-CMake-Documentation.rst:
```diff
@@ -0,0 +1,530 @@
+    find_package(Eigen3 REQUIRED)
+    ament_target_dependencies(my_library Eigen3)
+
+This will work very similar to the target version.
```
Maybe mention briefly what it does?

Also this approach ensures that include directories of multiple dependencies are ordered correctly in the light of overlay workspaces. For the target example above that is not the case hence it wouldn't work correctly for overlayed workspaces. Therefore I would lean to always suggest using `ament_target_dependencies` - in the future that function could be extended to also accept target names.

Martin-Idel replied:
I've never used overlay workspaces myself (I always used complete source builds for development), so I didn't know this, but that's an excellent reason to prefer `ament_target_dependencies`.

However
a) it would certainly be nice to support target names.
b) if the EXPORT command uses the same name as the target, I had trouble using `ament_target_dependencies`.

In addition, using `target_link_libraries` doesn't work very nicely with `ament_export_target_dependencies`, since you cannot use the `target_link_libraries(PUBLIC|INTERFACE|PRIVATE)` directive - you have to export all dependencies, even those which aren't referenced in public headers.

If you want, I can go back an gather information about the cases I encountered and open a ticket in `ament_cmake`. Maybe there is a way to support all cases and make `ament_target_dependencies` work cleanly in every scenario!

dirk-thomas commented on source/Tutorials/Ament-CMake-Documentation.rst:
```diff
@@ -0,0 +1,530 @@
+
+- The last large install command installs the library.
+  Archives and library files will be exported to the lib folder, runtime binaries will be installed to the bin folder and the path to installed headers is ``include``.
+  ``ament_cmake`` will make sure that the paths are adapted according to the mode of installation (merge-install or not)
```
I would suggest to remove this sentence since `ament_cmake` has no knowledge of this concept. I simply install the files to the `CMAKE_INSTALL_PREFIX`. The build tool is passing a different install prefix based on the `--merge-install` flag.

Martin-Idel replied:
Makes sense. I wanted to mention that the correct install paths are set by the build tool and users of `ament_cmake` don't have to worry about it, but I think this is not really the correct place.

dirk-thomas commented on source/Tutorials/Ament-CMake-Documentation.rst:
```diff
@@ -0,0 +1,530 @@
+Compiler and linker options
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+ROS 2 targets compilers which comply with the C++14 and C99 standard. Therefore it is customary to set the corresponding CMake flags:
```
I would suggest to qualify this statement for a specific ROS 2 distro (or multiple) since that is likely changing in newer distributions.

Martin-Idel replied:
Okay, then I'll mention that ROS2 targets these versions at least until `Crystal Clemmys` and that they may be updated later. Maybe it's best to include a link to https://index.ros.org/doc/ros2/Contributing/Developer-Guide/ where hopefully the targeted versions will be updated.

dirk-thomas replied:
Better reference the authoritive source of information: [REP 2000](http://www.ros.org/reps/rep-2000.html).

Martin-Idel replied:
Didn't know that. Thanks & done.

dirk-thomas commented on source/Tutorials/Ament-CMake-Documentation.rst:
```diff
@@ -0,0 +1,530 @@
+
+    find_package(ament_gtest REQUIRED)
+    ament_add_gtest(some_test <test_sources>
+      APPEND_ENV AMENT_PREFIX_PATH=${CMAKE_INSTALL_PREFIX})
```
This example doesn't make much sense since the tests might be invoked without installing the package. So relying on any artifacts to already be installed when running the test isn't reliable.

dirk-thomas commented on source/Tutorials/Ament-CMake-Documentation.rst:
```diff
@@ -0,0 +1,530 @@
+.. code-block:: cmake
+
+    ament_package(CONFIG_EXTRAS
+      ${CMAKE_SOURCE_DIR}/path/to/file.cmake
```
Is this (`CMAKE_SOURCE_DIR`) intentionally different from the above example?

Martin-Idel replied:
No, I copied these examples from some code patches in RViz and rosbag and didn't realize they weren't the same.

I'll harmonize the two versions and get rid of the variable alltogether since using relative paths is the simplest anyway.

dirk-thomas commented on source/Tutorials/Ament-CMake-Documentation.rst:
```diff
@@ -0,0 +1,530 @@
+.. code-block:: cmake
+
+    list(APPEND ${PROJECT_NAME}_CONFIG_EXTRAS
+      "${CMAKE_CURRENT_SOURCE_DIR}/path/to/file.cmake"
```
Relative paths are by default interpreted relative to `${CMAKE_CURRENT_SOURCE_DIR}`.

dirk-thomas commented on source/Tutorials/Ament-CMake-Documentation.rst:
```diff
@@ -0,0 +1,530 @@
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In addition to simple files with functions that can be used in other packages, you can also add extensions to ament.
+Those extensions are scripts which are executed with ``ament_package()``.
```
Those extensions are *not* necessarily executed with `ament_package()`. Each extension is triggered wherever the extension point is being invoked. For the below example `rosidl_generate_interfaces` that is e.g. when calling `rosidl_generate_interfaces` provided by the `rosidl_cmake` package.

dirk-thomas commented on source/Tutorials/Ament-CMake-Documentation.rst:
```diff
@@ -0,0 +1,530 @@
+Those extensions are scripts which are executed with ``ament_package()``.
+The most common use-case for ament extensions is probably registering rosidl message generators:
+When writing a generator, you normally want to generate all messages and services with your generator also without modifying the code for the message/service definition packages.
+This is possible by regisetring the generator as an extension to ``rosidl_generate_interfaces``.
```
Spelling: regisetring.

dirk-thomas commented on source/Tutorials/Ament-CMake-Documentation.rst:
```diff
@@ -0,0 +1,530 @@
+Especially when developing plugins or packages which allow plugins it is often essential to add resources to one ROS package from another (e.g. a plugin).
+Examples can be plugins for tools using the pluginlib.
+
+This can be achieved using the resource index within the ament index.
```
"ament index" and "resource index" are both a synonym for the same thing. We should probably have done a better job naming it consistently.

Martin-Idel replied:
Thanks, I always thought that the ament index would do some more stuff (maybe in the future) and the resource index was only supposed to be some subset. I tried to consistently name it "ament index" here and only mention once that this is sometimes called "resource index".

Martin-Idel commented:
@dirk-thomas Thank you for your input - I learned some new things. I reworked the document according to your suggestions and added some more information about extension points.

hidmic commented on source/Tutorials/Ament-CMake-Documentation.rst:
```diff
@@ -0,0 +1,572 @@
+
+The project setup is done by ``ament_package()`` and this call must occur exactly once per package.
+``ament_package()`` installs the ``package.xml``, registers the package with the ament index, and installs config (and possibly target) files for CMake so that it can be found by other packages using ``find_package``.
+Since ``ament_package()`` gathers a lot of information from the ``CMakeLists.txt`` it should be the last call in you CMakeLists.
```
@Martin-Idel `...in your CMakeLists.txt.`

hidmic commented on source/Tutorials/Ament-CMake-Documentation.rst:
```diff
@@ -0,0 +1,498 @@
+        $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
+        $<INSTALL_INTERFACE:include>)
+
+This adds all files in the folder ``${CMAKE_CURRENT_SOURCE_DIR}/include`` to the public interface during build time and
```
Also, I think there's an extra trailing newline here.

hidmic commented on source/Tutorials/Ament-CMake-Documentation.rst:
```diff
@@ -0,0 +1,572 @@
+      INCLUDES DESTINATION include
+    )
+
+Here, we assume that the folder ``includes`` contains the headers which need to be exported.
```
@Martin-Idel `include` in singular to match the CMake snippet.

hidmic commented on source/Tutorials/Ament-CMake-Documentation.rst:
```diff
@@ -0,0 +1,572 @@
+   Windows dlls are treated as runtime artifacts and installed into the ``RUNTIME DESTINATION`` folder.
+   It is therefore advised to not leave out the ``RUNTIME`` install even when developing libraries on Unix based systems.
+
+- Regarding the includes directory, the install command only adds information to CMake, it does not actually install the includes folder.
```
@Martin-Idel `includes directory/folder` -> ` ``include`` directory`, consistently.

hidmic commented on source/Tutorials/Ament-CMake-Documentation.rst:
```diff
@@ -0,0 +1,572 @@
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+If necessary, it is possible to query the ament index for resources via CMake.
+To do so, use there are three functions
```
@Martin-Idel `To do so, use there are three functions` -> `To do so, there are three functions:`

dirk-thomas commented on source/Tutorials/Ament-CMake-Documentation.rst:
```diff
@@ -0,0 +1,571 @@
+  For an example of when to use these arguments, see the discussion in `Adding resources`_.
+  For more information on how to use template files, see `the official documentation <https://cmake.org/cmake/help/v3.5/command/configure_file.html>`__.
+
+- ``CONFIG_EXTRAS_POST``: same as ``CONFIG_EXTRAS``, but the order in which the files are added differs: Those files will be added after files marked with ``CONFIG_EXTRAS`` are added.
```
The current text still doesn't clarify why `CONFIG_EXTRAS_POST` are not simply appended to `CONFIG_EXTRAS`. I would suggest the following:

> - `CONFIG_EXTRAS_POST`: same as `CONFIG_EXTRAS`, but the order in which the files are added differs.
> While `CONFIG_EXTRAS` files are included before the files generated for the `ament_export_*` calls the files from `CONFIG_EXTRAS_POST` are included afterwards.

Martin-Idel replied:
@dirk-thomas Thanks for your suggestion. I guess I didn't correctly understand the difference until now. Corrected.

dirk-thomas commented on source/Tutorials/Ament-CMake-Documentation.rst:
```diff
@@ -0,0 +1,572 @@
+      DESTINATION include
+    )
+
+    install(my_library
```
The first argument seems wrong. The signature of `install()` should start with `TARGETS`. See #318 for a patch.
